app/src/lib/editors/darwin.ts

```diff
@@ -0,0 +1,167 @@
+import * as Path from 'path'
+
+/**
+ * appPath will raise an error if it cannot find the program.
+ */
+const appPath: (bundleId: string) => Promise<string> = require('app-path')
+
+import {
+  LookupResult,
+  FoundEditor,
+  pathExists,
+  AtomLabel,
+  VisualStudioCodeLabel,
+  SublimeTextLabel,
+} from './shared'
+
+async function findAtomApplication(): Promise<LookupResult> {
+  const name = AtomLabel
+
+  try {
+    const installPath = await appPath('com.github.atom')
+    const path = Path.join(
+      installPath,
+      'Contents',
+      'Resources',
+      'app',
+      'atom.sh'
+    )
+    const exists = await pathExists(path)
+    if (!exists) {
+      log.debug(`Command line interface for ${name} not found at '${path}'`)
+      return {
+        name,
+        installed: true,
+        pathExists: false,
+      }
+    }
+    return {
+      name,
+      installed: true,
+      pathExists: true,
+      path,
+    }
+  } catch (error) {
+    log.debug(`Unable to locate ${name} installation`, error)
+    return { name, installed: false }
+  }
+}
+
+async function findCodeApplication(): Promise<LookupResult> {
+  const name = VisualStudioCodeLabel
+  try {
+    const installPath = await appPath('com.microsoft.VSCode')
+    const path = Path.join(
+      installPath,
+      'Contents',
+      'Resources',
+      'app',
+      'bin',
+      'code'
+    )
+    const exists = await pathExists(path)
+    if (!exists) {
+      log.debug(`Command line interface for ${name} not found at '${path}'`)
+      return {
+        name,
+        installed: true,
+        pathExists: false,
+      }
+    }
+    return {
+      name,
+      installed: true,
+      pathExists: true,
+      path,
+    }
+  } catch (error) {
+    log.debug(`Unable to locate ${name} installation`, error)
+    return { name, installed: false }
+  }
+}
+
+async function findSublimeTextApplication(): Promise<LookupResult> {
+  const name = SublimeTextLabel
+  try {
+    const sublimeApp = await appPath('com.sublimetext.3')
+    const path = Path.join(
+      sublimeApp,
+      'Contents',
+      'SharedSupport',
+      'bin',
+      'subl'
+    )
+
+    const exists = await pathExists(path)
```

Comment:
It kinda seems like this pattern of "get the path for the bundle ID, check for a subpath" is general enough to be extracted out so that we can 🔥 a lot of this duplicate code.